selectors/selector-syntax.md

```diff
@@ -0,0 +1,430 @@
+
+### Whitespace is Ignored
+
+Whitespace is completely ignored by the parser except for inside quoted strings.
+
+When extending this in the future, be aware that whitespace cannot be used as keyword boundaries (`"ab cd"` is identical to `"a bc d"`).
+We should have enough space for dozens of long and short names, but will want to write a tool to automatically look for ambiguities as well as improve developer experience with auto formatters and smart highlighters.
+
+### Parentheses are Usually Optional
+
+Parentheses annotate structure and are sometimes required for ambigious cases such as unions which contain an arbitrary number of selectors or selectors with optional parameters of conflicting types.
+
+However the parser can usually infer the structure without them because most selectors have a fixed or semi-fixed arity and certain types are only allowed at certain places.
+
+The best practice (and what the default formatting styles will enforce) is for human readable selectors to use parentheses liberally while URL embedding style will only contain the required ones.
+
+### Parameters can be Named
+
+Parameters can usually be inferred by their contextual position, but there are some cases where it's ambigious and needs to be specified.  There are more cases where it's good to annotate them for human clarity.
+
+For example, `recursive` has two required parameters and a 3rd optional one.
+
+```ts
+recursive(sequence: Selector, limit: int, stopAt?: Condition)
+```
+
+Written verbosely with parentheses, named parameters, and whitespace, it looks like this:
+
+```ipldsel
+recursive(
+    limit=5
+    sequence=...
+)
+```
+
+Depending on the context, we could omit the parentheses because the optional `stopAt` parameter is of type `Condition` and the parser likely expects something else after this node.
+
+Also we don't need to annotate `limit=` or `sequence=` since both are non-optional, and unique types.  Notice that the order doesn't matter and we can put `limit` before `sequence` because of unambigious types.
+
+Best practice is to annotate `limit`, but not `sequence` for human readable, and omit both for URL form.
+
+```ipldsel
+# Human Readable
+recursive(limit=5 ...)
+# URL Embeddable
+R(5...)
+```
+
+### Literal Values
+
+Some of the selectors accept literal values as parameters.  These are currently `String`, `{String:Selector}`, and `int`.
+
+#### Integers
+
+Integers can be encoded using common integer literals found in programming languages:
+
+```
+123 # Decimal
+0xfeed # Hex
+0o644 # Octal
+0b1001010 # Binary
+```
+
+Note these are case insensitive for both the hex digits as well as the base prefixes.
+
+#### Strings
+
+Strings are quoted using single quote, they can be escaped using double single quote.  You can include non url-safe characters between the quotes, but will need to escape the entire selector properly when embedding in a URL.
```

Comment:
Is `'` going to be robust enough if we wanted to start escaping more? Say we wanted to handle `/` characters in fields but not have a selector confuse it with a path, `'` won't help us here I think, it's the kind of thing you'd use `\` for: `parent/parent/a\/b` where `"a/b"` is a map key at some node.

Reply:
I don't understand the limitation.  It can escape any string you want by simply doubling any `'` characters found in the original string.  If you want to include `\/` in the string, it's fine.

Reply:
If you're talking about embedding the selector in a url component, then the forward slash is already escaped at the url embedding layer.  It's not included in the safe character set because `encodeURIComponent("foo/bar")` is `foo%2Fbar`

Reply:
I'm thinking about the case where we want _special_ characters, not just raw characters, in the same way that `\n` is a special character with many `\` escaping schemes. My point about `/` is hypothetical—imagine a case where we say that `'foo/bar/baz'` is parsed according to the rules you have here, it's a path shorthand. Right now, if you happen to have data with keys that contain the `/` character then paths aren't going to work nicely for you (bad luck, choose better keys buddy). But say we decide that it's important that we expand the range of characters that paths work for to include `/` characters _inside_ of keys, so we want to be able to switch between `/` for path separators and `/` as a character inside a key. A logical way of doing this is to escape it, turn it into something special that the parser has to treat differently ("oh, that's not a path separator because it's escaped, it must be part of the key string for this path segment"). How would you escape anything other than `'` when `'` is both your escaping character and your string terminator? If you have a key `b/a/z` at the end of your path, you can't: `'foo/bar/b'/a'/z'` because you've got termination confusion. If a character other than `'` was your escaping character then it's clearer: `'foo/bar/b\/a\/r'`.

Reply:
I see, you want the ability to embed things like newlines and tabs in the string without using actual newlines and tabs.  In the current scheme, it's quite possible to include them, but it's going to look weird.
```
'this is a string
with a newline'
```
vs
```
'this is a string\nwith a newline'
```

Reply:
I went ahead and switched to a more traditional string encoding.  It's now single quote only, but otherwise very similar to JSON with backslash escaping for whitespace, backslash and single quote.  I'm leaving unicode escaping out of this since unicode values can simply be included as-is and the JSON form of `\uxxxx` is only 16.